src/lib.rs

```diff
@@ -17,6 +17,48 @@ mod session;
 // Keep in sync with Cargo.toml.
 const RUSTLS_CRATE_VERSION: &str = "0.19.0";
 
+/// CastPtr represents the relationship between a snake case type (like rustls_client_session)
+/// and the corresponding Rust type (like ClientSession). For each matched pair of types, there
+/// should be an `impl CastPtr for snake_case { RustTy = SnakeCase }`.
+///
+/// This allows us to avoid using `as` in most places, and ensure that when we cast, we're
+/// preserving const-ness, and casting between the correct types.
+/// Implementing this is required in order to use `try_ref_from_ptr!`.
+pub(crate) trait CastPtr {
+    type RustTy;
+
+    fn cast_const_ptr(ptr: *const Self) -> *const Self::RustTy {
+        ptr as *const _
+    }
+
+    fn cast_mut_ptr(ptr: *mut Self) -> *mut Self::RustTy {
+        ptr as *mut _
+    }
+}
+
+/// Turn a raw const pointer into a reference. This is a generic function
+/// rather than part of the CastPtr trait because (a) const pointers can't act
+/// as "self" for trait methods, and (b) we want to rely on type inference
+/// against T (the cast-to type) rather than across F (the from type).
+pub(crate) fn try_from<F, T>(from: *const F) -> Option<&'static T>
```

Comment:
But that can't be expressed as long as the input parameter is not also a reference which provides the input lifetime, right?

Reply:
I think you are right. I tried rewriting this to use the lifetime elided from `try_from` ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bc96f5d1caa5cb94bfd592be494bb561), and it works, but I suspect that when you do `try_from<'a, F, T>(from: *const F) -> Option<&'a T>;`, `'a` winds up being `'static` anyway, so the compiler won't help you avoid unsafe references. I'm not sure how to do better than `'static` here.

Reply:
[The as_ref docs](https://doc.rust-lang.org/std/primitive.pointer.html#method.as_ref) use a template lifetime, and say "You must enforce Rust's aliasing rules, since the returned lifetime 'a is arbitrarily chosen and does not necessarily reflect the actual lifetime of the data. In particular, for the duration of this lifetime, the memory the pointer points to must not get mutated (except inside UnsafeCell)." I figure it's reasonable to do the same thing `as_ref` does, so I added the lifetime parameter.